homeassistant/components/here_travel_time/__init__.py

```diff
@@ -1 +1,172 @@
-"""The here_travel_time component."""
+"""The HERE Travel Time integration."""
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+import logging
+
+import async_timeout
+from herepy import NoRouteFoundError, RouteMode, RoutingApi, RoutingResponse
+
+from homeassistant.const import ATTR_ATTRIBUTION, CONF_UNIT_SYSTEM_IMPERIAL, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.location import find_coordinates
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt
+
+from .const import (
+    ATTR_DESTINATION,
+    ATTR_DESTINATION_NAME,
+    ATTR_DISTANCE,
+    ATTR_DURATION,
+    ATTR_DURATION_IN_TRAFFIC,
+    ATTR_ORIGIN,
+    ATTR_ORIGIN_NAME,
+    ATTR_ROUTE,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    NO_ROUTE_ERROR_MESSAGE,
+    TRAFFIC_MODE_ENABLED,
+    TRAVEL_MODES_VEHICLE,
+    HERERoutingData,
+    HERETravelTimeConfig,
+)
+
+PLATFORMS = [Platform.SENSOR]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HereTravelTimeDataUpdateCoordinator(DataUpdateCoordinator):
+    """HERETravelTime DataUpdateCoordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        api: RoutingApi,
+        config: HERETravelTimeConfig,
+    ) -> None:
+        """Initialize."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=DEFAULT_SCAN_INTERVAL),
+        )
+        self._api = api
+        self.config = config
+
+    async def _async_update_data(self) -> HERERoutingData | None:
+        """Get the latest data from the HERE Routing API."""
+        try:
+            async with async_timeout.timeout(10):
+                return await self.hass.async_add_executor_job(self._update)
+        except NoRouteFoundError as error:
+            raise UpdateFailed(NO_ROUTE_ERROR_MESSAGE) from error
+
+    def _update(self) -> HERERoutingData | None:
```

Comment:
This may be a matter of style/opinion, but i think this function is doing a bit too much beyond the scope of the actual DataUpdateCoordinator. It seems like the responsibilities of preparing the origin/destination can be separate, and the response parsing can be separate.

Reply:
I concur. I wanted to keep the changes as small as possible in this PR and submit code style changes afterwards.